File: Chapter 11/JS/section_11_4_secrets_management/secure/domain/secretsPort.js
/**
 * DOMAIN LAYER - Secrets Port (Interface)
 * 
 * This defines the contract for retrieving secrets. The application layer
 * depends on this abstraction, not on any specific secrets manager implementation.
 * 
 * ARCHITECTURAL NOTE:
 * This is a Port in Hexagonal Architecture. The application core defines
 * WHAT it needs (an interface), and the infrastructure layer provides
 * HOW it's implemented (AWS Secrets Manager, HashiCorp Vault, etc.).
 */

export class SecretsPort {
    /**
     * Retrieve database credentials from the secrets manager.
     * 
     * @abstract
     * @param {string} secretName - The name/ARN of the secret to retrieve
     * @returns {Promise<{username: string, password: string}>} Database credentials
     * @throws {Error} If the secret doesn't exist or IAM role lacks permission
     */
    async getDatabaseCredentials(secretName) {
        throw new Error('Not implemented');
    }
}

export class SecretNotFoundError extends Error {
    constructor(message) {
        super(message);
        this.name = 'SecretNotFoundError';
    }
}

export class PermissionError extends Error {
    constructor(message) {
        super(message);
        this.name = 'PermissionError';
    }
}

File: Chapter 11/Python/section_11_4_secrets_management/secure/domain/secrets_port.py
"""
DOMAIN LAYER - Secrets Port (Interface)

This defines the contract for retrieving secrets. The application layer
depends on this abstraction, not on any specific secrets manager implementation.
"""

from abc import ABC, abstractmethod
from typing import Tuple


class ISecretsProvider(ABC):
    """
    PORT: Abstract interface for secrets management.
    
    ARCHITECTURAL NOTE:
    This is a Port in Hexagonal Architecture. The application core defines
    WHAT it needs (an interface), and the infrastructure layer provides
    HOW it's implemented (AWS Secrets Manager, HashiCorp Vault, etc.).
    
    This allows us to:
    1. Swap secrets providers without changing business logic
    2. Test with mock implementations
    3. Support multiple cloud providers
    """
    
    @abstractmethod
    def get_database_credentials(self, secret_name: str) -> Tuple[str, str]:
        """
        Retrieve database credentials from the secrets manager.
        
        Args:
            secret_name: The name/ARN of the secret to retrieve
            
        Returns:
            Tuple of (username, password)
            
        Raises:
            SecretNotFoundError: If the secret doesn't exist
            SecretPermissionError: If the IAM role lacks permission to access the secret
        """
        pass


class SecretNotFoundError(Exception):
    """Raised when a requested secret cannot be found."""
    pass


class SecretPermissionError(Exception):
    """Raised when the application lacks IAM permission to access a secret."""
    pass
